src/main/java/com/iota/iri/service/milestone/MilestoneSolidifier.java

```diff
@@ -0,0 +1,247 @@
+     * Integer.MAX_VALUE.
+     *
+     * @return the Map.Entry holding the earliest milestone or a default Map.Entry(null, Integer.MAX_VALUE)
+     */
+    private Map.Entry<Hash, Integer> getEarliestUnsolidMilestoneEntry() {
+        try {
+            return Collections.min(unsolidMilestones.entrySet(), comparingInt(Map.Entry::getValue));
+        } catch (NoSuchElementException e) {
+            return new AbstractMap.SimpleEntry<>(null, Integer.MAX_VALUE);
+        }
+    }
+
+    /**
+     * This method removes the current earliest Milestone from the map and sets the internal pointers to the next
+     * earliest one.
+     *
+     * It is used to cycle through the unsolid milestones as they become solid or irrelevant for our node.
+     */
+    private void nextEarliestMilestone() {
+        unsolidMilestones.remove(earliestUnsolidMilestoneHash);
+
+        Map.Entry<Hash, Integer> nextEarliestMilestone = getEarliestUnsolidMilestoneEntry();
+
+        earliestUnsolidMilestoneHash = nextEarliestMilestone.getKey();
+        earliestUnsolidMilestoneIndex = nextEarliestMilestone.getValue();
+        earliestUnsolidMilestoneSolidificationAttempts = 0;
+    }
+
+    /**
+     * This method performs the actual solidity check on the selected milestone.
+     *
+     * It first checks if there is a milestone that has to be solidified and then issues the corresponding solidity
+     * check. In addition to issuing the solidity checks, it dumps a log message to keep the node operator informed
+     * about the progress of solidification.
+     *
+     * We limit the amount of transactions that may be processed during the solidity check, since we want to solidify
+     * from the oldest milestone to the newest one and not "block" the solidification with a very recent milestone that
+     * needs to traverse huge chunks of the tangle. If we fail to solidify a milestone for a certain amount of tries, we
+     * increase the amount of transactions that may be processed by using an exponential binary backoff strategy. The
+     * main goal of this is to give the solidification enough "resources" to discover the previous milestone (if it ever
+     * gets stuck because of a very long path to the previous milestone) while at the same time allowing fast solidity
+     * checks in "normal conditions".
+     *
+     * @return true if there are no unsolid milestones that have to be processed or if the earliest milestone is solid
+     */
+    private boolean isEarliestMilestoneSolid() {
+        if (earliestUnsolidMilestoneHash == null) {
+            return true;
+        }
+
+        if (unsolidMilestones.size() > 1) {
+            statusLogger.status("Solidifying milestone #" + earliestUnsolidMilestoneIndex + " [" + unsolidMilestones.size() + " left]");
+        }
+
+        try {
+            return transactionValidator.checkSolidity(
+            earliestUnsolidMilestoneHash,
+                true,
+                2 ^ Math.min(
```

Comment:
In Java the `^` operator is not power of, it is binary XOR! You have to use `Math.pow`.